exonum-java-binding/core/src/main/java/com/exonum/binding/core/blockchain/serialization/BlockSerializer.java

```diff
@@ -72,4 +86,60 @@ private static HashCode toHashCode(Hash hash) {
     ByteString bytes = hash.getData();
     return HashCode.fromBytes(bytes.toByteArray());
   }
+
+  private static Optional<HashCode> toOptionalHashCode(Hash hash) {
+    return Optional.of(hash)
+        .map(Hash::getData)
+        .filter(h -> !h.isEmpty())
+        .map(ByteString::toByteArray)
+        .map(HashCode::fromBytes);
+  }
+
+  private static Types.Hash toHashProtoOptional(Optional<HashCode> hash) {
+    ByteString bytes = hash
+        .map(HashCode::asBytes)
+        .map(ByteString::copyFrom)
+        .orElse(ByteString.EMPTY);
+
+    return Types.Hash.newBuilder()
+        .setData(bytes)
+        .build();
+  }
+
+  @VisibleForTesting
+  static ImmutableMap<String, ByteString> toHeadersMap(AdditionalHeaders headers) {
+    return headers.getHeaders().getEntryList()
+        .stream()
+        .collect(collectingAndThen(
+            toMap(
+                KeyValue::getKey,
+                KeyValue::getValue,
+                (o, n) -> {
+                  throw new IllegalStateException(String.format(
+                      "Should never happen. Duplicate key found in headers %s", headers));
+                },
+                LinkedHashMap::new),
+            ImmutableMap::copyOf)
+        );
+  }
+
+  @VisibleForTesting
+  static AdditionalHeaders toHeadersProto(ImmutableMap<String, ByteString> headers) {
+    return AdditionalHeaders.newBuilder()
+        .setHeaders(
+            KeyValueSequence.newBuilder()
+                .addAllEntry(
+                    headers.entrySet()
+                        .stream()
+                        .map(e -> KeyValue.newBuilder()
+                            .setKey(e.getKey())
+                            .setValue(e.getValue())
+                            .build())
+                        .collect(toList())
```

Comment:
Consider accumulating directly in the builder — that's what it is for — and not in the intermediate list.